rtt/os/macosx/fosi.h

```diff
@@ -237,24 +249,319 @@ extern "C"
     /* 		return -1; */
     /*     } */
 
-    // Mutex functions - support only those needed by TLSF
-	// opaque type
-    typedef struct rt_mutex_impl_t rt_mutex_impl_t;
-	// type created by tlsf.c (must work in C, not C++ code)
-    typedef rt_mutex_impl_t* rt_mutex_t;
-    int rtos_mutex_init(rt_mutex_t* m);
-    int rtos_mutex_destroy(rt_mutex_t* m);
-	int rtos_mutex_lock( rt_mutex_t* m);
-	int rtos_mutex_unlock( rt_mutex_t* m);
+    // Mutex functions
+
+
+    // Note: Apple's implementation of pthread_mutex_t does not have pthread_mutex_timedlock.
+    // Workaround: Pair the mutex with a condition variable that is signaled whenever the mutex is unlocked.
+    // This is also how boost::timed_mutex is implemented for the case BOOST_THREAD_USES_PTHREAD_TIMEDLOCK is
+    // not defined (cf. https://github.com/boostorg/thread/blob/boost-1.62.0/include/boost/thread/pthread/mutex.hpp#L181).
+    typedef struct {
+        pthread_mutex_t m;
+        pthread_cond_t cond;
+        bool is_locked;
+    } rt_mutex_t;
+    typedef struct {
+        pthread_mutex_t m;
+        pthread_cond_t cond;
+        bool is_locked;
+        pthread_t owner;
+        unsigned count;
+    } rt_rec_mutex_t;
+
+    #define _RTOS_INTERNAL_MUTEX_LOCK \
+      RTT_VERIFY(!(ret = pthread_mutex_lock((&(m->m))))); \
+      if (ret != 0) return ret
+
+    #define _RTOS_INTERNAL_MUTEX_UNLOCK_AND_RETURN \
+      RTT_VERIFY(!pthread_mutex_unlock(&(m->m))); \
+      return
+
+    static inline int rtos_mutex_init(rt_mutex_t* m)
+    {
+        pthread_mutexattr_t attr;
+        int ret = 0;
+
+        RTT_VERIFY(!(pthread_mutexattr_init(&attr)));
+        if (ret != 0) return ret;
+
+        // enable priority inheritance
+        // The mutex only protects the internal flags, so no use to enable priority inheritance.
+//        RTT_VERIFY(!(ret = pthread_mutexattr_setprotocol(&ma_t, PTHREAD_PRIO_INHERIT)));
+//        if (ret != 0) {
+//            RTT_VERIFY(!pthread_mutexattr_destroy(&attr));
+//            return ret;
+//        }
```

Comment:
Is keeping this comment useful for future reference? If so, consider reformatting :+1:

Reply:
Ideally the Mac OS X implementation of `rt_mutex_t` would also support priority inheritance like for the `gnulinux` target (in `rdt-next`) and pthread would support this, but because the pthread mutex only protects the `is_locked` flag it would be invalid to only enable `PTHREAD_PRIO_INHERIT` only because a thread waiting on the condition variable in [pthread_cond_wait](https://github.com/snrkiwi/rtt/pull/10/files/8109c52e1c8330e12ea63d4d669cbc9e34f2fa70#diff-0365fa9bbed2e68fc298c6a67d8d3b3bR341) would not inherit the priority of the thread that holds the mutex.

I can remove the commented code, but I would keep the text above.